delivery_carrier_label_default/tests/test_print_label.py

```diff
@@ -0,0 +1,130 @@
+    """Test label printing.
+
+    When running tests Odoo 11 renders PDF reports as HTML,
+    so we are going to test the shipping labels as HTML document only.
+    A good side effect: we are able to test the rendered content.
+    """
+    def check_label_content(self, b64_datas):
+        html_datas = base64.b64decode(b64_datas)
+        node = self.to_xml_node(html_datas)[0]
+        for div_class in ['page', 'address', 'recipient']:
+            tags = self.find_div_class(node, div_class)
+            self.assertEqual(len(tags), 1)
+
+    @patch_label_file_type
+    def test_print_default_label(self):
+        self.picking.generate_labels()
+        label = self.env['shipping.label'].search(
+            [('res_id', '=', self.picking.id)])
+        self.assertEquals(len(label), 1)
+        self.assertTrue(label.datas)
+        self.assertEquals(label.name, "Shipping Label.html")
+        self.assertEquals(label.file_type, 'html')
+        self.check_label_content(label.datas)
+
+    @patch_label_file_type
+    def test_print_default_label_multi_packs(self):
+        # create packs
+        self.picking.action_confirm()
+        self.picking.action_assign()
+        self.picking.move_line_ids[0].qty_done = 3
+        self.picking.move_line_ids[1].qty_done = 3
+        self.picking.put_in_pack()
+        for ope in self.picking.move_line_ids:
+            if ope.qty_done == 0:
+                ope.qty_done = 9
+                break
+        self.picking.put_in_pack()
+        self.picking.generate_labels()
+        label = self.env['shipping.label'].search(
+            [('res_id', '=', self.picking.id)])
+        self.assertEquals(len(label), 1)
+        self.assertTrue(label[0].datas)
+        self.assertEquals(label[0].name, "Shipping Label.html")
+        self.assertEquals(label[0].file_type, 'html')
+        self.check_label_content(label[0].datas)
+
+    @patch_label_file_type
+    def test_print_default_label_selected_packs(self):
+        # create packs
+        self.picking.action_confirm()
+        self.picking.action_assign()
+        self.picking.move_line_ids[0].qty_done = 3
+        self.picking.move_line_ids[1].qty_done = 3
+        self.picking.put_in_pack()
+        for ope in self.picking.move_line_ids:
+            if ope.qty_done == 0:
+                ope.qty_done = 9
+                break
+        self.picking.put_in_pack()
+        packs = self.picking.mapped(
+            'move_line_ids.result_package_id')
+        self.picking.generate_labels(package_ids=packs.ids)
+        labels = self.env['shipping.label'].search(
+            [('res_id', '=', self.picking.id)])
+        self.assertEquals(len(labels), 2)
+        for label in labels:
+            self.assertTrue(label.datas)
+            self.assertEquals(label.name, "Shipping Label.html")
+            self.assertEquals(label.file_type, 'html')
+            self.check_label_content(label.datas)
+
+    def setUp(self):
```

Comment:
pls, do not put tests set up at the bottom: they are the very beginning of a test suite...

Plus, pls, move the base class to SavepointCase and this code here to setUpClass otherwise you create all these records again again and again for each test.

Finally, use this trick at the beginning of setUpClass to turn off tracking:
```py
cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
```

Reply:
I agree, it was already at the bottom so I tend to reduce the diff size if possible ;) Np to move it.

Reply:
Done.